### PR TITLE
dequantizing splines

### PR DESCRIPTION
**Pull request by mo271**

### CODE REVIEW

File: jxl/src/features/spline.rs [hunk 1]
```diff
@@ -59,6 +63,21 @@ pub struct QuantizedSpline {
     pub sigma_dct: [i32; 32],
 }
 
+// Maximum number of spline control points per frame is
```
Comment (veluca93) [inline jxl/src/features/spline.rs:66]: Is this meant to be here?

Comment (mo271) [reply jxl/src/features/spline.rs:66]: we will use it soon, but right now it is not needed, deleted.

**Review by veluca93 — APPROVED** [review]
Can you confirm the numbers with @eustas and/or @sboukortt ?

**Comment by zond** [comment]
> Can you confirm the numbers with @eustas and/or @sboukortt ?

The numbers come directly from libjxl, I added tests that produce the same numbers, but I'll ask @sboukortt and @eustas to review.

**Review by eustas — APPROVED** [review]

File: jxl/src/util/test.rs [hunk 2]
```diff
@@ -25,6 +28,27 @@ macro_rules! assert_almost_eq {
...
+                    match $crate::util::test::abs_delta(left_val[index], right_val[index]).partial_cmp(max_error) {
+                        Some(std::cmp::Ordering::Greater) | None =>  panic!(
+                            "assertion failed: `(left ≈ right)`\n left: `{:?}`,\n right: `{:?}`,\n max_error: `{:?}`,\n left[{}]: `{}`,\n right[{}]: `{}`",
+                            left_val, right_val, max_error, index, left_val[index], index, right_val[index]
+                        ),
+                        _ => {}
+                    }
+                }
+            }
+        }
```
Comment (sboukortt) [inline jxl/src/util/test.rs:46]: Couldn’t this (outer) `match` be a `let`?

File: jxl/src/features/spline.rs [hunk 3]
```diff
@@ -108,6 +135,119 @@ impl QuantizedSpline {
...
+        result.estimated_area_reached = width_estimate * manhattan_distance;
+
+        // TODO: move this check to the outside, using the returned estimated_area_reached,
+        // making the caller keep track of the total estimated_area_readed
+        //if result.total_estimated_area_reached > area_limit {
+        //    return Err(Error::SplinesAreaTooLarge(
+        //        *total_estimated_area_reached,
+        //        area_limit,
+        //    ));
+        //}
```
[the remark this answers is not in the document]
Comment (zond) [reply jxl/src/features/spline.rs:247]: Yes - it's a piece of documentation about what we need to move :D

A later PR actually moves it and removes the TODO.

File: jxl/src/error.rs [hunk 4]
```diff
@@ -111,10 +111,16 @@ pub enum Error {
...
+    #[error("Too large area for spline: {0}, limit is {1}")]
+    SplinesAreaTooLarge(u64, u64),
+    #[error("Too large manhattan_distance reached: {0}, limit is {1}")]
+    SplinesDistanceTooLarge(u64, u64),
     #[error("Too many splines: {0}, limit is {1}")]
     SplinesTooMany(u32, u32),
     #[error("Too many control points for splines: {0}, limit is {1}")]
     SplinesTooManyControlPoints(u32, u32),
+    #[error("Spline point outside valid bounds: {0}, bounds are {1:?}")]
+    SplinesPointOutOfRange(i32, std::ops::Range<i32>),
```
[the remark this answers is not in the document]
Comment (zond) [reply jxl/src/error.rs]: Done

File: jxl/src/util/test.rs (hunk 2 above)
Comment (zond) [reply jxl/src/util/test.rs]: Done

**Review by sboukortt — APPROVED** [review]

**Pull request merged**